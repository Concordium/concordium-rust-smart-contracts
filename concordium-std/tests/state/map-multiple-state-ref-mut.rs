//! This test checks that you cannot have multiple `StateRefMut` from a
//! `StateMap` alive at the same time.
//!
//! When compiling it, the borrow-checker is supposed to throw an error.
use concordium_std::*;

pub fn main() {
    let mut state_builder = StateBuilder::open(ExternStateApi::open());
    let mut map: StateMap<u8, u8, _> = state_builder.new_map();
    map.insert(0, 1);
    map.insert(1, 2);
    // Get two mutable references and unwrap the options.
    let e1 = map.get_mut(&0u8).unwrap();
    let e2 = map.get_mut(&1u8).unwrap();
    // Use them, so we are certain that their lifetimes overlap.
    assert_eq!(*e1, *e2);
}
