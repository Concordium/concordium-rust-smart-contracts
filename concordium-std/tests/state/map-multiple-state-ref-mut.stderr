error[E0499]: cannot borrow `map` as mutable more than once at a time
  --> tests/state/map-multiple-state-ref-mut.rs:14:14
   |
13 |     let e1 = map.get_mut(&0u8).unwrap();
   |              --- first mutable borrow occurs here
14 |     let e2 = map.get_mut(&1u8).unwrap();
   |              ^^^ second mutable borrow occurs here
15 |     // Use them, so we are certain that their lifetimes overlap.
16 |     assert_eq!(*e1, *e2);
   |                 -- first borrow later used here
